### ISSUE TITLE
Add an option to output atomic css

### ISSUE THREAD

**Issue by asvny** [callstack/linaria#225]
**Do you want to request a *feature* or report a *bug*?**
Feature

**What is the current behavior?**
-

**If the current behavior is a bug, please provide the steps to reproduce and a minimal repository on GitHub that we can `yarn install` and `yarn test`.**

**What is the expected behavior?**
The current behavior is

```js
// Input 

const header = css`
  text-transform: uppercase;
  font-family: ${fonts.heading};
  font-size: ${modularScale(2)};
`;

// Current Output
const header = 'header_t1ugh8t9';

// Expected Output
const header = 'tt_swer3 ff_Sers34 fs_5tsdf';
/*
tt_swer3  -> text-transform
ff_Sers34 -> font-family
fs_5tsdf -> font-size
*/
```
**Please provide your exact Babel configuration and mention your Linaria, Node, Yarn/npm version and operating system.**

**Comment by asvny** [comment]
@satya164 I very excited about this and if you are fine with this feature, I am ready to experiment by integrating https://github.com/linkedin/opticss with linaria to optimize the output.

**Note: This can be added as non-breaking change by adding this as feature flag until it gets stable.**

**Comment by satya164** [comment]
@asvny I think it'll be great. maybe we would want to have a separate package or entry point `linaria/atomic` if the implementation diverges too much. The only problem I see is that this feature can only work with `css` tags. With `styled` tags, there will be specificity issues when decorating existing styled component.

**Comment by asvny** [comment]
@satya164 I am not able to get this point
 ` With styled tags, there will be specificity issues when decorating existing styled component.`

Can you please provide an example?

**Comment by satya164** [comment]
@asvny so for example, say you have this:

```js
const Button = styled.button`
  background: orange;
`;

const FancyButton = styled(Button)`
  background: blue;
`;
```

here, the background of `FancyButton` will override the background of `Button` because we generate a more specific selector (`Button` will generate something like `.abcdef` and `FancyButton` will generate something like `.abcdef.tyuio`). could be difficult with atomic class names. but probably we can avoid atomic class names for components like `FancyButton` too properly handle it.

**Comment by giuseppeg** [comment]
it is possible to resolve the styles if you use a specific naming convention and process the atomic class names right to left when applying them. I do that in DSS https://survivejs.com/blog/dss-interview/ and in https://github.com/giuseppeg/style-sheet

**Comment by jayu** [comment]
I'm closing this issue because it's pretty old and not upvoted, we have other priorities right now

**Comment by LucaColonnello** [comment]
I would have loved to see this coming though! So sad to see it has been closed!
This is the approach taken by the new Facebook style lib, which would be open source at some point.
But I think, considering it seems that lib is quite opinionated in the way to write CSS (ReactNative style), it would be good to see this in Linaria as an opt-in feature.

If I have some time, does anyone mind me spiking it?

**Comment by LucaColonnello** [comment]
@giuseppeg I'm impressed with your work on this, and as you are waay ahead of me, any chance we could have a brief conversation about how to possibly spike this in Linaria as a POC?
(It would still be Callstack final say if they are happy to integrate this or not and how)

**Comment by thymikee** [comment]
@jayu mind reopening then? :) I think being able to have atomic css would be great, because it scales well in larger codebases.

**Comment by Jayphen** [comment]
I'm a big fan of this proposal — and as Stylex will likely be open-sourced at some point, it makes sense to be ahead of the ball on implementing this.

The great benefit is that it results in logarithmic instead of linear growth of CSS when adding new features with styling. This can result in massive savings for large apps (in Facebook's case it reduced their CSS payload by 80%)

**Comment by LucaColonnello** [comment]
Well said @jayu and @Jayphen!
There's also an element of bringing the same innovation that company like Facebook are discovering and promoting outside of those companies tools, to be part of open source alternatives that have a different model.

I'm personally a big big fan of Linaria, as I thing THAT is how CSS-in-JS should be done.
But there's also a less opinionated view on top of this, which is the one that CSS-in-JS with a JS runtime does not suit every use case (i.e. a Gatsby blog or other types of mostly static pages with few interactions). The idea of using CSS-in-JS just as a Dev Experience tool as opposed to pay the cost in the final result as well, is something that brings us closer to how the browser uses the resources.
Plus, compile time, I mean, I don't think can be beaten...

That's why I think that generated Atomic CSS is a good combination with this as it fits the paradigm of Linaria very well, where you write CSS the way you learned it, so this is just an optimisation on top...

**Comment by jayu** [comment]
Good to see some discussion in this repo :D I'm not opossed to the feature, I was doing issue triage a month ago and I closed this since it wasn't upvoted and I didn't get the idea behind it. 
I still don't have time to implement it myself, but I'm open to check any PoC! This is not easy thing to implement in linaria, but if someone will manage to do it, I'm good. 
I'm a bit afraid about build time, but we will see :D

**Comment by LucaColonnello** [comment]
Hey @jayu!
Definitely I'd like to make a PoC of this and see how it goes.

It will be challenging for sure, that's why I might just make a PoC of the final outcome and open a PR to discuss all the point where we need to improve it, maybe with a roadmap...

**Comment by jayu** [comment]
I looked into how atomic CSS might work and unfortunately, due to we support nesting in linaria, it would be impossible to have both atomic output and nesting classnames. We could probably have atomic CSS as a derivation or however it could be called, with restricted syntax support. Atomic CSS would make it very hard to make https://github.com/callstack/linaria/issues/244 work. 

I'm also wondering how we could deal with redundantly generated class names. 
The problem is we process each file with babel, but the next file does not about the previous processing, so we may end up with a declaration for a particular class name repeated 10 times. 
We could implement some cache that will save the information about whether the class was already generated or not, how to decide then for which file the class should be generated. What if the class will be removed from the file, but needed by other files. We will need a different build configuration for production and development then. It will be definitely challenging :D 

StyleX will probably have some super fancy compiler for that, we are not that advanced :P

**Comment by asvny** [comment]
My two cents ...

StyleX is quite simple and follows `react-native` stylesheet pattern. It is authored like
```js
const styles = stylex.create({
 backgroundColor: 'red',
 color: 'white'
})
```
I possibly assume that they force developers to author only longhand properties and think in terms of logical properties like `margin-inline, block-size` etc..,

Then babel compiles it
```js
const styles = {
 backgroundColor: 'hash1',
 color: 'hash2'
}
```

When it needs to resolve multiple object styles, it does something similar to `Object.assign`

```js
const styles1 = {
 backgroundColor: 'hash1',
 color: 'hash2'
}

const styles2 = {
 backgroundColor: 'hash3',
 textTransform: 'hash4'
}

const classes = Object.values(Object.assign({}, styles1, styles2));
```
So coming back to `Linaria`, it is hard to get it right because of the nesting problem. There are two things that we can do
1. Disallow nesting and throw an error during dev time.
2. Compile nesting to atomic CSS, (personally I think it is a bad option) 
Example: cxs css-in-js

The reason why I suggested opticss is that for efficient atomic CSS generation, we have to know the template/component on how it is being used and based on that we can allow certain combinator selectors (like the ones you see in LinkedIn's CSS Blocks). OptiCSS has ways to solve cascade by analyzing the document/jsx etc...(https://github.com/linkedin/opticss/blob/master/packages/resolve-cascade/README.md). And I think at the sample time, OptiCSS is experimental and not yet ready for production usage.

It definitely is a hard problem to solve, given the case nesting is allowed.

**Comment by giuseppeg** [comment]
fwiw [style-sheet](https://github.com/giuseppeg/style-sheet) my (pof) CSS in JS library is based on the same principles of stylex, actually I started to build it way before FB talked about it publicly. 

To extract styles it uses Linaria's evaluation helper, taken from Linaria's Babel plugin. It's just hasn't been updated in a while https://github.com/giuseppeg/style-sheet/pull/17

**Comment by LucaColonnello** [comment]
I can see the challenges in this using babel @jayu .

I was thinking to use webpack for it instead, but yeah not sure if feasible.
On @giuseppeg's comment, I wonder if it's applicable without using objects, using simple strings as Linaria does (which is mainly why I like Linaria, as it's easier to migrate to and keeps syntax aligned and compatible with standard CSS)...

**Comment by LucaColonnello** [comment]
For what's worth, on @asvny's comment, https://github.com/CraigCav/css-zero :)

**Comment by johanholmerin** [comment]
If you are interested in stylex, I have created an open-source implementation of it [here](https://github.com/johanholmerin/style9).

**Comment by CanRau** [comment]
Very interesting conversation 😍
Wanted to mention [Fela](http://fela.js.org/) which produces atomic CSS like this:

```css
.a { text-align: center }
.b { padding: 5px 10px }
.c { font-size: 14pt }
.d { border-radius: 5px }
.e:hover { font-size: 16pt  #}
```
❤️🚀

**Comment by jpnelson** [comment]
This is a really interesting discussion – I've done a bit of a prototyping into what would be required for supporting this in Linaria – it'd be a fairly different implementation, but we could keep a similar API shape (not too dissimilar from [css-zero](https://github.com/callstack/linaria/issues/225#issuecomment-628049099). At a high level, what I'd suggest would be something like this in React:

```jsx
import { css, styled } from '@linaria/atomic';

const blue = css`
   color: blue;
`;

const Component = styled.div`
  color: blue;
  border: 1px solid black;
`;

export default function () {
  return <Component>
    <p className="blue">Hello</p>
  </Component>
}
```

which would produce:

```html
<div class="atm_foo123">
  <p class="atm_foo123 atm_bar456">
    Hello
  </p>
</div>
```
```css
.atm_foo123 {
  color: blue;
}
.atm_bar456 {
  border: 1px solid black;
}
```

To get there we'd need:

* A `@linaria/atomic` package, that contains the new API (note: we'd probably also want new stylelint rules to enforce the fact that nesting / cascading isn't allowed in atomic css)
* Support in the babel plugin for the atomic logic in this new API
* The `collect` utility from `@linaria/server` to support sorting / deduping these atomic rules (I think prefixing them with a consistent prefix, eg. `atm_` could help here)

For @jayu 's point about [deduping the produced rules](https://github.com/callstack/linaria/issues/225#issuecomment-627477526), I think we dedupe rules at a *file* level, and rely on the bundler to dedupe those as well in produced css files (so, add webpack support, or similar).

For us, the big advantage of doing this is that the atomic rules don't depend on declaration order – we can determine which atomic rules to apply at build time. Additionally, the total css size can be reduced.

It'd be a fairly big new addition to the library though – @jayu, if I created a proof of concept along the lines of the above, would you be open to discussing the potential inclusion? I think it'll end up being a bunch of new work though so it'd be great to hear what your thoughts are for the direction on this going forward.

**Comment by Anber** [comment]
Hi @jpnelson!

The separate `@linaria/atomic` sounds cool. We will definitely face a lot of challenges but let's try! :)

**Comment by aquelehugo** [comment]
I think stylex (Facebook CSS-in-JS library to be released) approach should be taken into consideration before implementing that: https://www.youtube.com/watch?v=ur-sGzUWId4

**Comment by jpnelson** [comment]
Definitely agree! That's a source of inspiration for that feature – another example of how it can be done is https://github.com/atlassian-labs/compiled . I have a work in progress on this, I'll put up a PR today to start getting feedback on it 🎉

**Comment by jpnelson** [comment]
@Anber @jayu I had a first pass on it here – https://github.com/callstack/linaria/pull/867 . Feedback would be helpful!

**Comment by juanferreras** [comment]
@jpnelson just wanted to chime in and say the `@linaria/atomic` package works excellently! Congrats and many thanks to everyone involved! linaria with opt-in atomic css support is really such a fantastic combination 🎉 

Is the idea still to support `styled` (similarly to `css`, probably from `@linaria/atomic` to make it opt-in)? Or has your work thus far found any blockers on that? thanks!

**Comment by juanferreras** [comment]
@Anber @jpnelson building upon your work I've went ahead and created a draft PR #966 to support `import { styled } from @linaria/atomic` too! I'd love to get your feedback on it (including, of course, if you think it's not a good idea to support it at all).

I've created it as a draft because there are a few open questions on the description (regarding dynamic interpolations and `styled(OtherComponent)`) that I'd love to hear thoughts on.

thanks so much!

**Comment by mlenser** [comment]
This is awesome, thanks so much for the work on this @juanferreras!

We're looking to use Linaria for our organization's upcoming design system and having this would be great!

Is there somewhere to hear more about the release cycles of Linaria or when we might be able to expect the next one which would support this? Thanks!

**Comment by Anber** [comment]
@mlenser I planned to release it an hour ago but found a small problem with TS. Hopefully, I will fix it and release it later today.

**Comment by Anber** [comment]
Has been released in `3.0.0-beta.19`.
Please note, that this release contains a lot of changes under the hood and despite the fact that all tests are green and my thousand-components-project works fine, I don't sure that it will work as stable as `beta.18`. If you find any problems, please let me know as soon as possible.
